OpenRA.Mods.D2k/Traits/Sandworm.cs

```diff
@@ -0,0 +1,130 @@
+#region Copyright & License Information
+/*
+* Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+* This file is part of OpenRA, which is free software. It is made
+* available to you under the terms of the GNU General Public License
+* as published by the Free Software Foundation. For more information,
+* see COPYING.
+*/
+#endregion
+
+using System;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.D2k.Traits
+{
+	class SandwormInfo : WandersInfo, Requires<MobileInfo>, Requires<RenderUnitInfo>, Requires<AttackBaseInfo>
+	{
+		[Desc("Time between rescanning for targets (in ticks).")]
+		public readonly int TargetRescanIntervalInTicks = 32;
+
+		[Desc("The radius in which the worm \"searches\" for targets.")]
+		public readonly int MaxSearchRadius = 32;
+
+		[Desc("The range at which the worm launches an attack regardless of noise levels.")]
+		public readonly int LaunchAttackRange = 3;
+
+		public readonly string IdleSequenceName = "idle";
+
+		public override object Create(ActorInitializer init) { return new Sandworm(init.Self, this); }
+	}
+
+	class Sandworm : Wanders, ITick, INotifyBecomingIdle
+	{
+		readonly Lazy<Mobile> mobile;
+		readonly Lazy<RenderUnit> renderUnit;
+		readonly Lazy<AttackBase> attackTrait;
+		readonly SandwormInfo info;
+
+		public bool IsAttacking;
+
+		int targetCountdown;
+
+		public Sandworm(Actor self, SandwormInfo info)
+			: base(self, info)
+		{
+			this.info = info;
+			mobile = Exts.Lazy(self.Trait<Mobile>);
```

Comment:
Again, these traits aren't guaranteed to exist on this actor (you are just hoping here).